hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/FetchStyleLoadPlanBuildingAssociationVisitationStrategy.java

```diff
@@ -190,11 +200,107 @@ protected FetchStrategy adjustJoinFetchIfNeeded(
 			return new FetchStrategy( fetchStrategy.getTiming(), FetchStyle.SELECT );
 		}
 
-		if ( attributeDefinition.getType().isCollectionType() && isTooManyCollections() ) {
-			// todo : have this revert to batch or subselect fetching once "sql gen redesign" is in place
-			return new FetchStrategy( fetchStrategy.getTiming(), FetchStyle.SELECT );
+		final FetchSource currentSource = currentSource();
+		final Type attributeType = attributeDefinition.getType();
+
+
+		if ( attributeType.isCollectionType() ) {
+			if ( isTooManyCollections() ) {
+				// todo : have this revert to batch or subselect fetching once "sql gen redesign" is in place
+				return new FetchStrategy( fetchStrategy.getTiming(), FetchStyle.SELECT );
+			}
+			if ( currentSource.resolveEntityReference() != null ) {
+				CollectionPersister collectionPersister =
+						(CollectionPersister) attributeDefinition.getType().getAssociatedJoinable( sessionFactory() );
+				// Check if this is an eager "mappedBy" (inverse) side of a bidirectional
+				// one-to-many/many-to-one association, with the many-to-one side
+				// being the associated entity's ID as in:
+				//
+				//  @Entity
+				//  public class Foo {
+				//      ...
+				//      @OneToMany(mappedBy = "foo", fetch = FetchType.EAGER)
+				//      private Set<Bar> bars = new HashSet<>();
+				//  }
+				//  @Entity
+				//  public class Bar implements Serializable {
+				//      @Id
+				//      @ManyToOne(fetch = FetchType.EAGER)
+				//      private Foo foo;
+				//      ...
+				//  }
+				//
+				if ( fetchStrategy.getTiming() == FetchTiming.IMMEDIATE &&
+						fetchStrategy.getStyle() == FetchStyle.JOIN &&
+						collectionPersister.isOneToMany() &&
+						collectionPersister.isInverse() ) {
+					// This is an eager "mappedBy" (inverse) side of a bidirectional
+					// one-to-many/many-to-one association
+					final EntityType elementType = (EntityType) collectionPersister.getElementType();
+					final Type elementIdType = ( (EntityPersister) elementType.getAssociatedJoinable( sessionFactory() ) ).getIdentifierType();
+					if ( elementIdType.isComponentType() && ( (CompositeType) elementIdType ).isEmbedded() ) {
+						final EmbeddedComponentType elementIdTypeEmbedded = (EmbeddedComponentType) elementIdType;
+						if ( elementIdTypeEmbedded.getSubtypes().length == 1 &&
+								elementIdTypeEmbedded.getPropertyNames()[ 0 ].equals( collectionPersister.getMappedByProperty() ) ) {
+							// The associated entity's ID is the other (many-to-one) side of the association.
```

Comment:
Could we simplify this comment a bit?  Something like

```
// The owning side of the inverse collection is defined by the associated entity's id.
// 
// Because of how Loaders process ids when processing the ResultSet, this condition would 
// lead to an infinite loop.  Adjust the fetch to use a SELECT fetch instead of JOIN to avoid
```

?